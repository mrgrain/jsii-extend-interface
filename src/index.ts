export * from './jsii-struct-builder';
export * from './jsii-struct-file';
export * from './typescript-renderer';
