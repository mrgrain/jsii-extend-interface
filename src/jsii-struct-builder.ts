import { InterfaceType, Property } from '@jsii/spec';

/**
 * Build a jsii struct
 */
export class JsiiStructBuilder {
  private targetSpec: InterfaceType;
  private properties: Map<string, Property>;

  public constructor(base: InterfaceType) {
    this.targetSpec = base;
    this.properties = new Map(base.properties?.map((p) => [p.name, p]));
  }

  /**
   * Omit this prop
   */
  public omit(prop: string): JsiiStructBuilder {
    this.properties.delete(prop);

    return this;
  }

  /**
   * Add a prop
   */
  public add(prop: Property): JsiiStructBuilder {
    this.properties.set(prop.name, prop);

    return this;
  }

  /**
   * Update an existing prop
   */
  public update(name: string, update: Partial<Property>): JsiiStructBuilder {
    const old = this.properties.get(name);

    if (!old) {
      throw `Unable top update property '${name}' in '${this.targetSpec.fqn}: Property does not exists, please use add'`;
    }

    const updated = {
      ...old,
      ...update,
      docs: {
        ...old?.docs,
        ...update?.docs,
        custom: {
          ...old?.docs?.custom,
          ...update?.docs?.custom,
        },
      },
    };

    return this.add(updated);
  }

  /**
   * Get the current state of the builder
   */
  public get(): InterfaceType {
    this.targetSpec.properties = Array.from(this.properties.values());

    return this.targetSpec;
  }
}
