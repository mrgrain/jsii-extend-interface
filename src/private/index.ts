export * from './assembly';
export * from './utils';
