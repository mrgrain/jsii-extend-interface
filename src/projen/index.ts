export * from './projen-struct';
export * from './ts-interface';
