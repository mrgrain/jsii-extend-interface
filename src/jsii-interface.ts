import { dirname, join, posix } from 'path';
import { InterfaceType, Property, TypeKind } from '@jsii/spec';
import { Component, typescript } from 'projen';
import { findInterface } from './assembly';
import { InterfaceFile } from './interface-file';

export interface JsiiInterfaceOptions {
  /**
   * The name of the interface
   */
  readonly name: string;
  /**
   * Doc string for the interface
   * @default - Interface name
   */
  readonly description?: string;
  /**
   * The fqn of the interface
   *
   * Used to auto-add imports.
   * All referenced types are loaded based on the fqn hierarchy.
   *
   * See `importLocations` for customization options.
   *
   * @default `${project.name}.${options.name}`
   */
  readonly fqn?: string;
  /**
   * Output file path
   * @default - inside `srcdir` or `outdir` with the fqn turned into a path
   */
  readonly filePath?: string;
  /**
   * The module locations assemblies should be imported from
   *
   * All types are imported from the top-level of the module.
   * This map can be used to overwrite the default import locations.
   *
   * For local imports, the import traverse up a number of levels equivalent to the number of fqn parts of the rendered type.
   * E.g. if the the rendered type is `pkg.interface`, it is assumed to be at the top-level and imports from the same assembly would be from `./`.
   * If the rendered type is `pkg.nested.sub.interface` a local import will be from `../../`.
   *
   * @default - uses the assembly name for external packages
   * local imports traverse up a number of levels equivalent to the number of fqn parts
   */
  importLocations?: Record<string, string>;
  /**
   * The properties of this interface
   *
   * When extending, these will replace existing properties.
   *
   * @default []
   */
  readonly props?: {
    [name: string]: Omit<Property, 'name'>;
  };
  /**
   * Extends this jsii interface
   */
  readonly extends?: string;
  /**
   * Omit these properties from the parent interface
   */
  readonly omitProps?: string[];
  /**
   * Update these properties from the parent interface
   *
   * Property definitions are merged non-recursively.
   * I.e. when updating `type` a complete new type must be provided
   */
  readonly updateProps?: {
    [name: string]: Partial<Property>;
  };
}

/**
 * A component generating a jsii-compatible interface
 */
export class JsiiInterface extends Component {
  public constructor(
    project: typescript.TypeScriptProject,
    options: JsiiInterfaceOptions
  ) {
    super(project);

    const properties = Object.entries(options.props || {}).map(([name, p]) => ({
      name,
      ...p,
    }));

    const fqn = options.fqn ?? `${project.name}.${options.name}`;
    const assembly = fqn.split('.').at(0) ?? project.name;
    const targetSpec: InterfaceType = {
      kind: TypeKind.Interface,
      assembly,
      fqn,
      name: options.name,
      docs: {
        summary: options.description ?? options.name,
      },
      properties,
    };

    if (options.extends) {
      const sourceSpec = findInterface(options.extends, true);
      const omit = [
        ...(options.omitProps || []),
        ...properties.map((p) => p.name),
      ];
      targetSpec.properties!.push(
        ...(sourceSpec.properties || [])
          .map((p) => {
            if (options.updateProps?.[p.name]) {
              return {
                ...p,
                ...options.updateProps?.[p.name],
                docs: {
                  ...p.docs,
                  ...options.updateProps?.[p.name].docs,
                  custom: {
                    ...p.docs?.custom,
                    ...options.updateProps?.[p.name].docs?.custom,
                  },
                },
              };
            }
            return p;
          })
          .filter((p) => !omit.includes(p.name))
      );
    }

    targetSpec.properties = targetSpec.properties?.sort((a, b) =>
      a.name.localeCompare(b.name)
    );

    const baseDir = project.srcdir ?? project.outdir;
    const outputFile = options.filePath ?? join(baseDir, fqnToPath(fqn));
    new InterfaceFile(project, outputFile, targetSpec, {
      importLocations: {
        [assembly]: relativeImport(outputFile, baseDir),
        ...options.importLocations,
      },
    });
  }
}
function relativeImport(from: string, target: string): string {
  const diff = posix.relative(dirname(from), target);
  if (!diff) {
    return '.' + posix.sep;
  }
  return diff + posix.sep;
}

function fqnToPath(fqn: string): string {
  return join(...fqn.split('.').slice(1)) + '.ts';
}
