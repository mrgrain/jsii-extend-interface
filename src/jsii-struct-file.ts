import { InterfaceType } from '@jsii/spec';
import { Project, SourceCodeOptions, TextFile } from 'projen';
import {
  TypeScriptRenderer,
  TypeScriptRendererOptions,
} from './typescript-renderer';

/**
 * Options for `JsiiStructFile`.
 */
export interface JsiiStructFileOptions
  extends TypeScriptRendererOptions,
    SourceCodeOptions {}

/**
 * A Jsii Struct rendered as TypeScript interface
 */
export class JsiiStructFile extends TextFile {
  public constructor(
    project: Project,
    filePath: string,
    spec: InterfaceType,
    options: JsiiStructFileOptions = {}
  ) {
    super(project, filePath, options);

    const renderer = new TypeScriptRenderer(options);
    this.addLine(`// ${this.marker}`);
    this.addLine(renderer.renderStruct(spec));
  }
}
