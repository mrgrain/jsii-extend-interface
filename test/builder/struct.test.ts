import { Struct, TypeScriptRenderer } from '../../src';

test('can render a struct directly', () => {
  // ARRANGE
  const renderer = new TypeScriptRenderer();

  // ACT
  const struct = Struct.empty('@my-scope/my-pkg.MyFunctionProps');
  struct.mixin(Struct.fromFqn('projen.typescript.ProjenrcOptions'));

  // PREPARE
  const renderedFile = renderer.renderStruct(struct);

  // ASSERT
  expect(renderedFile).toMatchSnapshot();
  expect(renderedFile).toContain('projenCodeDir');
  expect(renderedFile).toContain('filename');
});
