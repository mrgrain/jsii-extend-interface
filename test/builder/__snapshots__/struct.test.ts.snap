// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can render a struct directly 1`] = `
"
export interface MyFunctionProps {
  /**
   * Whether to use \`SWC\` for ts-node.
   * @default false
   * @stability experimental
   */
  readonly swc?: boolean;
  /**
   * A directory tree that may contain *.ts files that can be referenced from your projenrc typescript file.
   * @default "projenrc"
   * @stability experimental
   */
  readonly projenCodeDir?: string;
  /**
   * The name of the projenrc file.
   * @default ".projenrc.ts"
   * @stability experimental
   */
  readonly filename?: string;
}
"
`;
